frame/nomination-pools/src/lib.rs

```diff
@@ -0,0 +1,2005 @@
+			ensure!(!Delegators::<T>::contains_key(&who), Error::<T>::AccountBelongsToOtherPool);
+
+			let mut bonded_pool = BondedPool::<T>::get(pool_id).ok_or(Error::<T>::PoolNotFound)?;
+			bonded_pool.ok_to_join()?;
+
+			// We just need its total earnings at this point in time, but we don't need to write it
+			// because we are not adjusting its points (all other values can calculated virtual).
+			let reward_pool = RewardPool::<T>::get_and_update(pool_id)
+				.defensive_ok_or_else(|| Error::<T>::RewardPoolNotFound)?;
+
+			bonded_pool.try_inc_delegators()?;
+			let points_issued = bonded_pool.try_bond_funds(&who, amount, BondType::Later)?;
+
+			Delegators::insert(
+				who.clone(),
+				Delegator::<T> {
+					pool_id,
+					points: points_issued,
+					// At best the reward pool has the rewards up through the previous era. If the
+					// delegator joins prior to the snapshot they will benefit from the rewards of
+					// the active era despite not contributing to the pool's vote weight. If they
+					// join after the snapshot is taken they will benefit from the rewards of the
+					// next 2 eras because their vote weight will not be counted until the
+					// snapshot in active era + 1.
+					reward_pool_total_earnings: reward_pool.total_earnings,
+					unbonding_era: None,
+				},
+			);
+			bonded_pool.put();
+			Self::deposit_event(Event::<T>::Bonded {
+				delegator: who,
+				pool_id,
+				bonded: amount,
+				joined: true,
+			});
+
+			Ok(())
+		}
+
+		/// Bond `extra` more funds from `origin` into the pool to which they already belong.
+		///
+		/// Additional funds can come from either the free balance of the account, of from the
+		/// accumulated rewards, see [`BondExtra`].
+		// NOTE: this transaction is implemented with the sole purpose of readability and
+		// correctness, not optimization. We read/write several storage items multiple times instead
+		// of just once, in the spirit reusing code.
+		#[pallet::weight(
+			T::WeightInfo::bond_extra_transfer()
+			.max(T::WeightInfo::bond_extra_reward())
+		)]
+		#[transactional]
+		pub fn bond_extra(origin: OriginFor<T>, extra: BondExtra<BalanceOf<T>>) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+			let (mut delegator, mut bonded_pool, mut reward_pool) =
+				Self::get_delegator_with_pools(&who)?;
+
+			let (points_issued, bonded) = match extra {
+				BondExtra::FreeBalance(amount) =>
+					(bonded_pool.try_bond_funds(&who, amount, BondType::Later)?, amount),
+				BondExtra::Rewards => {
+					let claimed = Self::do_reward_payout(
+						&who,
+						&mut delegator,
+						&mut bonded_pool,
+						&mut reward_pool,
+					)?;
+					(bonded_pool.try_bond_funds(&who, claimed, BondType::Later)?, claimed)
+				},
+			};
+			delegator.points = delegator.points.saturating_add(points_issued);
+
+			Self::deposit_event(Event::<T>::Bonded {
+				delegator: who.clone(),
+				pool_id: delegator.pool_id,
+				bonded,
+				joined: false,
```

Comment:
Why did you go for a bool here instead of `BondType`?

Reply:
I assume its easier for clients reading events